doc/sphinxext/missing_references.py

```diff
@@ -0,0 +1,69 @@
+"""
+This is a sphinx extension to freeze your broken reference problems
+when using ``nitpicky = True``.
+
+The basic operation is:
+
+1. Add this extension to your ``conf.py`` extensions.
+2. Add ``missing_references_freeze = True`` to your ``conf.py``
+3. Run sphinx-build. It will generate ``missing-references.json``
+    next to your ``conf.py``.
+4. Remove ``missing_references_freeze = True`` from your
+    ``conf.py`` (or set it to ``False``)
+5. Run sphinx-build again, and ``nitpick_ignore`` will
+    contain all of the previously failed references.
+
+"""
+
+import json
+import os.path
+
+
+def record_missing_reference_handler(app, env, node, contnode):
+    if not app.config.missing_references_freeze:
+        # no-op when we are disabled.
+        return
+
+    if not hasattr(env, "missing_reference_record"):
+        env.missing_reference_record = []
+    record = env.missing_reference_record
+
+    domain = node["refdomain"]
+    typ = node["reftype"]
+    target = node["reftarget"]
+
+    dtype = "{}:{}".format(domain, typ)
+
+    record.append([dtype, target])
```

Comment:
I don't know if the information is available at that point, but if possible, it would be nice to also record the filename and the line number.